File: sections/28_dp_i_mixed_contest/problems/a_training_schedule/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (!(cin >> n)) return 0;
    const int INF = 1e9;
    array<int, 3> dp = {0, INF, INF};
    for (int day = 0; day < n; day++) {
        int available;
        cin >> available;
        array<int, 3> ndp = {min({dp[0], dp[1], dp[2]}) + 1, INF, INF};
        if (available & 1) ndp[1] = min(dp[0], dp[2]);
        if (available & 2) ndp[2] = min(dp[0], dp[1]);
        dp = ndp;
    }
    cout << min({dp[0], dp[1], dp[2]}) << '\n';
    return 0;
}

File: sections/28_dp_i_mixed_contest/problems/a_training_schedule/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n = data[0]
    days = data[1 : 1 + n]
    inf = 10**9
    dp = [0, inf, inf]  # rest, C++, Python as the previous day's action
    for available in days:
        ndp = [min(dp) + 1, inf, inf]
        if available & 1:
            ndp[1] = min(dp[0], dp[2])
        if available & 2:
            ndp[2] = min(dp[0], dp[1])
        dp = ndp
    print(min(dp))


if __name__ == "__main__":
    main()
